torch/csrc/api/include/torch/optim/serialize.h

```diff
@@ -40,6 +40,129 @@ void serialize(
+
+    torch::Tensor size_tensor;
+    param_group_archive.read("params/size", size_tensor);
+    const size_t size = size_tensor.item<int64_t>();
+    std::vector<std::string> params;
+    for (size_t index = 0; index < size; ++index) {
+      IValue ivalue;
+      param_group_archive.read(
+        "params/" + c10::to_string(index), ivalue);
+      std::string element = ivalue.toStringRef();
+      params.push_back(element);
+    }
+    serialize::InputArchive param_group_options_archive;
+    param_group_archive.read("options", param_group_options_archive);
+    DerivedOptimizerParamOptions param_group_options(0);
+    param_group_options.serialize(param_group_options_archive);
+    param_groups.push_back(std::make_pair(params, c10::guts::make_unique<DerivedOptimizerParamOptions>(param_group_options)));
+  }
+}
+
+// Utility function to save state and param_groups
+template <typename DerivedOptimizerParamState, typename DerivedOptimizerParamOptions>
+void serialize(
+    serialize::OutputArchive& archive,
+    const detail::OptimizerBase& optimizer) {
+
+  serialize::OutputArchive state_archive(archive.compilation_unit());
+  serialize<DerivedOptimizerParamState>(state_archive, optimizer.state_);
+  archive.write("state", state_archive);
+
+  serialize::OutputArchive param_groups_archive(archive.compilation_unit());
+  serialize<DerivedOptimizerParamOptions>(param_groups_archive, optimizer.param_groups_);
+  archive.write("param_groups", param_groups_archive);
+}
+
+// Utility function to load state and param_groups and update state
+template <typename DerivedOptimizerParamState, typename DerivedOptimizerParamOptions>
+void serialize(
+    serialize::InputArchive& archive,
+    detail::OptimizerBase& optimizer) {
+
+  serialize::InputArchive state_archive;
+  archive.read("state", state_archive);
+  ska::flat_hash_map<std::string, std::unique_ptr<OptimizerParamState>> state;
+  serialize<DerivedOptimizerParamState>(state_archive, state);
+
+  serialize::InputArchive param_groups_archive;
+  archive.read("param_groups", param_groups_archive);
+  std::vector<std::pair<std::vector<std::string>, std::unique_ptr<OptimizerOptions>>> param_groups;
+  serialize<DerivedOptimizerParamOptions>(param_groups_archive, param_groups);
+
+  // update state
+  TORCH_CHECK(param_groups.size() == optimizer.param_groups_.size(), "loaded state dict has a different number of parameter groups");
+  for(size_t i=0; i<param_groups.size(); i++) {
+    std::vector<std::string> saved_group_keys = param_groups[i].first;
+    std::vector<Tensor> params = optimizer.param_groups_[i].params();
+    TORCH_CHECK(saved_group_keys.size() == params.size(), "loaded state dict contains a parameter group that doesn't match the size of optimizer's group");
+    for(size_t idx = 0; idx<params.size(); idx++) {
+      optimizer.state_[c10::guts::to_string(params[idx].unsafeGetTensorImpl())]=std::move(state[saved_group_keys[idx]]);
+    }
+  }
+}
```

Comment:
We should add tests to make sure the full state of the optimizer (including `state` and `param_groups`) can be serialized and deserialized correctly. ;)  (the current test in `TEST(SerializeTest, Optim_Adagrad)` doesn't make the value checks for `state` and `param_groups` very explicit.)

Reply:
Umm will the way to check if state is being deserialized properly is run the python function, save the output, store the values in a file and then load them to compare with our output?

Reply:
I was thinking if it's okay to add overload functions for == operator in class OptimizerOptions, ParamGroups, OptimizerState and make them virtual so that each optimizer can define the function for its options and state. This will be helpful in writing the Optim serialize test